homeassistant/components/media_player/cambridgeaudio.py

```diff
@@ -0,0 +1,382 @@
+"""
+Support for Cambridge Audio Network Audio Players (StreamMagic platform).
+
+Example configuration.yaml entry:
+media_player:
+    platform: cambridgeaudio
+    host: 192.168.x.y
+    command_off: "OFF"
+    name: "CA 851N"
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.cambridgeaudio/
+"""
+import logging
+# pylint: disable=unused-import
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    SUPPORT_TURN_OFF, SUPPORT_TURN_ON,
+    SUPPORT_PAUSE, SUPPORT_PLAY, SUPPORT_PREVIOUS_TRACK, SUPPORT_NEXT_TRACK,
+    SUPPORT_STOP, SUPPORT_SEEK, SUPPORT_SHUFFLE_SET,
+    SUPPORT_SELECT_SOURCE, SUPPORT_CLEAR_PLAYLIST,
+    SUPPORT_VOLUME_SET, SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_STEP,
+    MediaPlayerDevice, PLATFORM_SCHEMA, MEDIA_TYPE_MUSIC)
+
+from homeassistant.const import (STATE_ON, STATE_OFF, STATE_UNKNOWN,
+                                 STATE_PLAYING, STATE_PAUSED, STATE_IDLE,
+                                 CONF_HOST, CONF_NAME, CONF_COMMAND_OFF)
+
+import homeassistant.helpers.config_validation as cv
+import homeassistant.util.dt as dt_util
+
+REQUIREMENTS = ['stream_magic==0.16']
+DOMAIN = 'cambridgeaudio'
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'Cambridge Audio Streamer'
+CONF_SOURCES = 'sources'
+KNOWN_HOSTS = []
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_COMMAND_OFF, default='OFF'): cv.string
+})
+
+# supported features in all configurations
+SUPPORT_CAMBRIDGE_COMMON = SUPPORT_TURN_OFF | SUPPORT_TURN_ON |\
+                           SUPPORT_PLAY | SUPPORT_SELECT_SOURCE |\
+                           SUPPORT_CLEAR_PLAYLIST | SUPPORT_STOP
+
+# supported features when device is configured as pre-amp
+SUPPORT_CAMBRIDGE_PREAMP = SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET |\
+                           SUPPORT_VOLUME_STEP
+
+# supported features for file-based media (i.e. not internet streams)
+SUPPORT_CAMBRIDGE_FILEMEDIA = SUPPORT_SEEK | SUPPORT_PAUSE |\
+                              SUPPORT_NEXT_TRACK | SUPPORT_PREVIOUS_TRACK |\
+                              SUPPORT_SHUFFLE_SET
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Cambridge Audio platform."""
+    from stream_magic import device as cadevice
+    from stream_magic import discovery as ca
+
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+
+    poweroff_command = config.get(CONF_COMMAND_OFF, 'OFF')
+
+    hosts = []
+    sm = ca.StreamMagic()   # pylint: disable=C0103
+
+    # found a new host that was specified in config file
+    if CONF_HOST in config and host not in KNOWN_HOSTS:
```

Comment:
This check should just check if discovery_info is None. Then we know we should try to set up via config. Then, after that, check if the configured host is already among known hosts and abort if so.